File: Shortest XY distance in Grid.py
class Solution:
    def shortestXYDist(self, grid, N, M):
        dist = [[float('inf')] * M for _ in range(N)]
        
        # check top and left
        for y in range(N):
            for x in range(M):
                if grid[y][x] == 'X':
                    dist[y][x] = 0
                else:
                    if y > 0:
                        dist[y][x] = min(dist[y][x], dist[y - 1][x] + 1)
                    if x > 0:
                        dist[y][x] = min(dist[y][x], dist[y][x - 1] + 1)
        
        # check bottom and right
        for y in range(N - 1, -1, -1):
            for x in range(M - 1, -1, -1):
                if grid[y][x] == 'X':
                    dist[y][x] = 0
                else:
                    if y < N - 1:
                        dist[y][x] = min(dist[y][x], dist[y + 1][x] + 1)
                    if x < M - 1:
                        dist[y][x] = min(dist[y][x], dist[y][x + 1] + 1)
        
        # return mininum distance
        ans = 1e9
        
        for i in range(N):
            for j in range(M):
                if grid[i][j] == 'Y':
                    ans = min(ans, dist[i][j])
        
        return ans

File: Shortest XY distance in Grid.cpp
class Solution {
  public:
    int shortestXYDist(vector<vector<char>> grid, int N, int M) {
        int dist[N][M];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                dist[i][j] = 1e9;
            }
        }

        // check top and left
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (grid[i][j] == 'X')
                    dist[i][j] = 0;
                else {
                    if (i > 0) dist[i][j] = min(dist[i][j], dist[i - 1][j] + 1);
                    if (j > 0) dist[i][j] = min(dist[i][j], dist[i][j - 1] + 1);
                }
            }
        }

        // check bottom and right
        for (int i = N - 1; i >= 0; i--) {
            for (int j = M - 1; j >= 0; j--) {
                if (grid[i][j] == 'X')
                    dist[i][j] = 0;
                else {
                    if (i < N - 1)
                        dist[i][j] = min(dist[i][j], dist[i + 1][j] + 1);
                    if (j < M - 1)
                        dist[i][j] = min(dist[i][j], dist[i][j + 1] + 1);
                }
            }
        }

        int ans = 1e9;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (grid[i][j] == 'Y') ans = min(ans, dist[i][j]);
            }
        }

        return ans;
    }
};
